File: neetcode/1. Two Sum/1.java
/*
Method 1.0:
Intuition: use a map to keep track of the element and its index, check if target - num is present
in map, then return indices
TC: O(N)
SC: O(N)
 */

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            if (map.containsKey(complement)) {
                return new int[]{map.get(complement), i};
            }

            map.put(nums[i], i);
        }

        return new int[]{-1, -1};
    }
}

File: neetcode/1. Two Sum/1.kt
/*
Method 1.0:
Intuition: use a map to keep track of the element and its index, check if target - num is present
in map, then return indices
TC: O(N)
SC: O(N)
 */

class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val map = HashMap<Int, Int>()

        for (i in nums.indices) {
            val complement = target - nums[i]
            if (map.containsKey(complement)) {
                return intArrayOf(map[complement]!!, i)
            }

            map[nums[i]] = i
        }

        return intArrayOf(-1, -1)
    }
}
